compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,132 +75,280 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
-        if (succeeded(tile) && *tile == tileOp) bdIdCount++;
+        if (succeeded(tile)) addDmaToBatch(*tile, dmaOp);
       }
+      if (!currDmaOp) currDmaOp = dmaOp;
     } else if (auto npuWaitOp = dyn_cast<AMDAIE::NpuDmaWaitOp>(op)) {
       for (AMDAIE::NpuDmaCpyNdOp npuDmaOp : npuWaitOp.getDmaOps()) {
-        // Reached the DMA wait operation, stop counting.
-        if (npuDmaOp == currDmaOp) return;
-      }
-    } else if (auto forOp = dyn_cast<scf::ForOp>(op)) {
-      uint32_t subLoopBdIdCount = 0;
-      getNumRequiredBdIds(forOp, nullptr, tileOp, shimTileToGeneratorMap,
-                          subLoopBdIdCount);
-      std::optional<uint32_t> subIterations = getNumberIterations(forOp);
-      if (subIterations) {
-        bdIdCount += subLoopBdIdCount * subIterations.value();
-      } else {
-        bdIdCount += subLoopBdIdCount;
+        // Reached the DMA wait operation, reset tracking of current DMA op for
+        // the tile.
+        if (npuDmaOp == currDmaOp) {
+          currDmaOp = nullptr;
+          if (npuDmaOp.getSource()) {
+            FailureOr<AMDAIE::TileOp> tile =
+                getGeneratorTileOp<CopyOpOperateOn::Source>(
+                    npuDmaOp, shimTileToGeneratorMap);
+            if (succeeded(tile)) updateCurrentTileBatch(*tile);
+          }
+          if (npuDmaOp.getTarget()) {
+            FailureOr<AMDAIE::TileOp> tile =
+                getGeneratorTileOp<CopyOpOperateOn::Target>(
+                    npuDmaOp, shimTileToGeneratorMap);
+            if (succeeded(tile)) updateCurrentTileBatch(*tile);
+          }
+        }
       }
     }
   }
+  return tileDmaBatchGraph;
 }
 
-/// Creates a BD ID operation for the given DMA copy operation.
-/// If the DMA copy operation is inside a loop, the BD ID operation will be
-/// created with a semi-affine expression to assign different BD IDs for each
-/// iteration. Otherwise, a constant BD ID will be assigned.
-template <CopyOpOperateOn OperateOn>
-FailureOr<AMDAIE::BdIdOp> getBdIdOp(
-    IRRewriter &rewriter, AMDAIE::NpuDmaCpyNdOp &npuDmaOp,
-    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap,
-    DenseMap<AMDAIE::BdIdOp, SmallVector<uint32_t>> &bdIdOpToBdIdsMap) {
-  // Get the TileOp.
-  FailureOr<AMDAIE::TileOp> maybeTileOp =
-      getGeneratorTileOp<OperateOn>(npuDmaOp, shimTileToGeneratorMap);
-  if (failed(maybeTileOp)) return failure();
-  AMDAIE::TileOp tileOp = maybeTileOp.value();
+/// Given a non-empty DmaBatch find the immediate surrounding parent op.
+static Operation *findParentOpOfBatch(DmaBatch *dmaBatch) {
+  if (!dmaBatch->currentDmaOps.empty())
+    return dmaBatch->currentDmaOps[0]->getParentOp();
+  assert(!dmaBatch->immediateInnerBatches.empty() &&
+         "DmaBatch found with no DmaOps and no immediate inner batches");
+  Operation *parentOp = findParentOpOfBatch(dmaBatch->immediateInnerBatches[0]);
+  assert(parentOp && "Found inner DmaBatch with no parent");
+  return parentOp->getParentOp();
+}
```

Comment:
This should be a struct/class method imo as it only operates on members in `DmaBatch`.